### PR TITLE
Timeout connection requests when in a deadlock.

### LINKED ISSUE #63: Detect connection pool deadlocks
Something like this will hang the client, and in db log we get

> LOG:  unexpected EOF on client connection with an open transaction

```swift
Client.find(id, on: req.db).flatMap {
    /// Validate we want to make updates
    req.db.transaction { db in
        // Do stuff
```

I don't really want to create a transaction if I don't have to. In the case where I discovered this I'm doing multiple queries against multiple tables, and only under certain conditions do I then create a transaction and update multiple different tables.

### PR DESCRIPTION
**Pull request by MrLotU**
Fixes a bug when requesting `n+1` connections from a connection pool with limit `n` where each new connection waited on the previous one.

Previously, this would create a deadlock. Now, a timeout can be added to `EventLoopConnectionPool`, that will return a failed future if the timeout is reached. Timeout defaults to 10 seconds but can be configured.
Fixes #63

### CODE REVIEW

File: Sources/AsyncKit/ConnectionPool/ConnectionPoolError.swift [hunk 1]
```diff
@@ -2,4 +2,6 @@
 public enum ConnectionPoolError: Error {
     /// The connection pool has shutdown.
     case shutdown
+    /// The connection creationg timed out.
```
Comment (MrMage) [inline Sources/AsyncKit/ConnectionPool/ConnectionPoolError.swift:5]: typo: creationg

File: Sources/AsyncKit/ConnectionPool/ConnectionPoolError.swift [hunk 2]
```diff
@@ -2,4 +2,6 @@
...
+    case connectionCreateTimeout
```
Comment (MrMage) [inline Sources/AsyncKit/ConnectionPool/ConnectionPoolError.swift:6]: Technically, this is not "creating", but rather "requested". Maybe rephrase it?

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift [hunk 3]
```diff
@@ -197,9 +204,19 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
             // connections are exhausted, we must wait for one to be returned
             logger.debug("Connection pool exhausted on this event loop, adding request to waitlist")
             let promise = eventLoop.makePromise(of: Source.Connection.self)
-            self.waiters.append((logger, promise))
+            let uuid = UUID()
+            self.waiters.append((logger, promise, uuid))
+            
+            let task = eventLoop.scheduleTask(in: self.creationTimeout) { [weak self, logger, promise] in
+                logger.error("Connection creation timed out. This might indicate a connection deadlock in your application.")
```
Comment (MrMage) [inline Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:211]: Same here — rename it to requesting?

File: Tests/AsyncKitTests/ConnectionPoolTests.swift [hunk 4]
```diff
@@ -134,6 +134,30 @@ final class ConnectionPoolTests: XCTestCase {
...
+            maxConnections: 1,
+            creationTimeout: .seconds(1),
+            on: self.eventLoopGroup.next()
+        )
+        defer { try! pool.close().wait() }
+        _ = pool.requestConnection()
+        let start = Date()
+        let a = pool.requestConnection()
+        do {
+            _ = try a.wait()
```
Comment (MrMage) [inline Tests/AsyncKitTests/ConnectionPoolTests.swift:152]: Could we use `XCTAssertThrowsError` instead of a try-catch block?

File: Tests/AsyncKitTests/ConnectionPoolTests.swift [hunk 5]
```diff
@@ -134,6 +134,30 @@ final class ConnectionPoolTests: XCTestCase {
...
         }
     }
+    
+    // https://github.com/vapor/async-kit/issues/63
+    func testDeadlock() {
+        let foo = FooDatabase()
+        let pool = EventLoopConnectionPool(
+            source: foo,
+            maxConnections: 1,
+            creationTimeout: .seconds(1),
```
Comment (MrMage) [inline Tests/AsyncKitTests/ConnectionPoolTests.swift:144]: Could this be reduced to e.g. ~100ms to improve test runtimes? Or, even better, is there an event loop group for testings that lets us "speed up time" for testing?

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift [hunk 6]
```diff
@@ -197,9 +204,19 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
...
+                if let idx = self?.waiters.firstIndex(where: { $0.2 == uuid }) {
```
Comment (MrMage) [inline Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:212]: Both the search and the remove would have linear runtime each. That is not ideal, but hopefully this occurs rarely enough to not be a bottleneck.

Unless in those 10 seconds before timing out so many waiters accumulate that it actually does become a bottleneck... When stuff fails, it often all fails at once, after all.

File: Sources/AsyncKit/ConnectionPool/ConnectionPoolError.swift (hunk 2 above)
Comment (tanner0101) [inline Sources/AsyncKit/ConnectionPool/ConnectionPoolError.swift:6]: We can't add new cases to this enum unfortunately. I'd recommend a new struct or enum.

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift [hunk 7]
```diff
@@ -36,7 +41,7 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
     
     /// All requests for a connection that were unable to be fulfilled
     /// due to max connection limit having been reached.
-    private var waiters: CircularBuffer<(Logger, EventLoopPromise<Source.Connection>)>
+    private var waiters: CircularBuffer<(Logger, EventLoopPromise<Source.Connection>, UUID)>
```
Comment (tanner0101) [inline Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:44]: ELPromise.value is a reference type, so you can use `===` to compare instead of storing a UUID.

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift [hunk 8]
```diff
@@ -60,11 +65,13 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
     public init(
         source: Source,
         maxConnections: Int,
+        creationTimeout: TimeAmount = .seconds(10),
```
Comment (tanner0101) [inline Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:68]: I'd call this `requestTimeout`.

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift [hunk 9]
```diff
@@ -197,9 +204,19 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
             // connections are exhausted, we must wait for one to be returned
             logger.debug("Connection pool exhausted on this event loop, adding request to waitlist")
             let promise = eventLoop.makePromise(of: Source.Connection.self)
-            self.waiters.append((logger, promise))
+            let uuid = UUID()
+            self.waiters.append((logger, promise, uuid))
+            
+            let task = eventLoop.scheduleTask(in: self.creationTimeout) { [weak self, logger, promise] in
```
Comment (MrMage) [inline Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:210]: Why do `logger` and `promise` need to be captured by the block?

Comment (MrLotU) [reply Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift]: `logger` is needed for the `.error` logging. `promise` is used to fail the promise if the timeout is over & (after latest push) for the `===` comparison.

Comment (MrMage) [reply Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift]: Yes, but why can't those be captured implicitly?

Comment (MrLotU) [reply Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift]: Good question. Not sure why I did this in this way 🤔. I can remove the explicit capture if preferred. CC @tanner0101

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift [hunk 10]
```diff
@@ -55,16 +60,20 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
...
     ///     - maxConnections: Limits the number of connections that can be open.
     ///                       Defaults to 1.
+    ///     - requestTimeout: Timeout for requesting a new connection.
+    ///                       Defaults to 10 seconds.
     ///     - logger: For lifecycle logs.
     ///     - on: Event loop.
     public init(
         source: Source,
         maxConnections: Int,
+        requestTimeout: TimeAmount = .seconds(10),
```
Comment (tanner0101) [inline Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:70]: FWIW, Redis's connection pool is using a timeout of 60 seconds: https://gitlab.com/Mordil/RediStack/-/merge_requests/116/diffs#8fde653c2095c02534d7072fa60cdb002ac223c5_0_208

The docs note: 
> for maximum compatibility with the existing implementation, we use a fairly-long timeout:

cc @lukasa

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift [hunk 11]
```diff
@@ -198,8 +207,16 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
             logger.debug("Connection pool exhausted on this event loop, adding request to waitlist")
             let promise = eventLoop.makePromise(of: Source.Connection.self)
             self.waiters.append((logger, promise))
-            // return waiter
-            return promise.futureResult
+            
+            let task = eventLoop.scheduleTask(in: self.requestTimeout) { [weak self, logger, promise] in
+                logger.error("Connection request timed out. This might indicate a connection deadlock in your application.")
+                if let idx = self?.waiters.firstIndex(where: { let p = $0.1; return p.futureResult === promise.futureResult }) {
```
Comment (tanner0101) [inline Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:213]: This could be nice:
```swift
guard let self = self else { return }
```

Comment (tanner0101) [inline Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:213]: I think closure labels and some whitespace here would be nice to avoid `let p = $0.1;` which is overly terse.

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift [hunk 12]
```diff
@@ -198,8 +207,16 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
...
+                    self?.waiters.remove(at: idx)
+                }
+                promise.fail(ConnectionPoolTimeoutError.connectionRequestTimeout)
+            }
+            
+            return promise.futureResult.always { _ in task.cancel() }
```
Comment (tanner0101) [inline Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:219]: Is cancelling a task that has already completed fine?

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift (hunk 10 above)
Comment (Lukasa) [reply Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:70]: Yeah, we were very defensive there. Maybe more than we should have been.

File: Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift (hunk 12 above)
Comment (MrLotU) [reply Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift:219]: I did a quick test where the task would always finish before cancelling and didn't see anything going wrong, so guessing it's fine. Not 100% sure though.

cc @Lukasa

**Comment by tanner0101** [comment]
These changes are now available in [1.2.0](https://github.com/vapor/async-kit/releases/tag/1.2.0)
